reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java

```diff
@@ -51,17 +56,32 @@ static void parse(Class<?> moduleClass, @Nullable Object instance,
         }
 
         if (method.getAnnotation(IntoSet.class) != null) {
+          wrapper = TypeWrapper.SET;
           throw notImplemented("@IntoSet");
         }
         if (method.getAnnotation(ElementsIntoSet.class) != null) {
+          wrapper = TypeWrapper.SET;
           throw notImplemented("@ElementsIntoSet");
         }
         if (method.getAnnotation(IntoMap.class) != null) {
+          wrapper = TypeWrapper.MAP;
           throw notImplemented("@IntoMap");
         }
 
         Annotation[] annotations = method.getAnnotations();
-        Type returnType = method.getGenericReturnType();
+        Type returnType;
+        switch (wrapper) {
+          case NONE:
+            returnType = method.getGenericReturnType();
+            break;
+          case OPTIONAL:
+            Type genericReturnType = method.getGenericReturnType();
+            returnType = new TypeUtil.ParameterizedTypeImpl(null, Optional.class,
+                genericReturnType);
+            break;
+          default:
+            throw notImplemented(wrapper.toString());
```

Comment:
unreachable because of above `notImplemented` calls, did you want to remove those?